### PR TITLE
prevent render canvas with quality less than 100%

### PR DESCRIPTION
**Pull request by vitonsky**
Fix issue https://github.com/fabricjs/fabric.js/issues/7502

### CODE REVIEW

File: src/static_canvas.class.js [hunk 1]
```diff
@@ -232,7 +232,7 @@
      * @private
      */
     _isRetinaScaling: function() {
-      return (fabric.devicePixelRatio !== 1 && this.enableRetinaScaling);
+      return (fabric.devicePixelRatio > 1 && this.enableRetinaScaling);
```
Comment (asturur) [inline src/static_canvas.class.js:235]: you also need to change getRetinaScaling to return a number that is at minimum 1 i think.

Comment (vitonsky) [reply src/static_canvas.class.js:235]: Done

Comment (Dok11) [reply src/static_canvas.class.js:235]: Maybe make sense to switch places of these parts for performance? Checking a boolean variable should be faster than comparing two values. Like this: `this.enableRetinaScaling && fabric.devicePixelRatio > 1`.

Comment (vitonsky) [reply src/static_canvas.class.js:235]: @Dok11 i don't sure. JS engine probably is don't know whats variable contain, so will not check variable as boolean directly. But symbol `>` may suggest that it's just math operation which is instant

Comment (Dok11) [reply src/static_canvas.class.js:235]: @vitonsky I even tested it, and it seems like it doesn't care :D
https://jsbench.me/zkkxq3rjv0/1

image

**Comment by ShaMan123** [comment]
@vitonsky could you push a commit that calculates retina scaling for within [`_onResize`](https://github.com/fabricjs/fabric.js/blob/7efc732d8a050d743a66c89d8b1c4e5f942191f4/src/mixins/canvas_events.mixin.js#L372) as suggested in https://github.com/fabricjs/fabric.js/issues/7502#issuecomment-986287369

File: src/static_canvas.class.js (hunk 1 above)
Comment (asturur) [reply src/static_canvas.class.js:235]: i m sure performance is not an issue here, but in the JS bench seems like Math.random and toFixed are actually the 2 things that will matter for performance. no?

**Review by asturur — APPROVED** [review]

**Comment by asturur** [comment]
we broke so many tests. not sure why it wasn't visible in the pr

File: src/static_canvas.class.js [hunk 2]
```diff
@@ -232,15 +232,15 @@
...
     },
 
     /**
      * @private
      * @return {Number} retinaScaling if applied, otherwise 1;
      */
     getRetinaScaling: function() {
-      return this._isRetinaScaling() ? fabric.devicePixelRatio : 1;
+      return Math.max(1, fabric.devicePixelRatio);
```
Comment (asturur) [inline src/static_canvas.class.js:243]: ok this was the error.
We lessen the importance of the boolean here.
If retina scaling is set as false, whatever is the devicePixelRatio, we return 1.
My bad for not noticing it.

Comment (vitonsky) [reply src/static_canvas.class.js:243]: Probably we can fix it like this?
```js
this._isRetinaScaling() ? Math.max(1, fabric.devicePixelRatio) : 1
```

**Comment by vitonsky** [comment]
@ShaMan123 
> @vitonsky could you push a commit that calculates retina scaling from within [`_onResize`](https://github.com/fabricjs/fabric.js/blob/7efc732d8a050d743a66c89d8b1c4e5f942191f4/src/mixins/canvas_events.mixin.js#L372) as suggested in [#7502 (comment)](https://github.com/fabricjs/fabric.js/issues/7502#issuecomment-986287369)

I'm not sure that we can do it on the library level. If you wish, make own pull request with this change if possible

**Pull request merged**